### PR TITLE
Ensure that rational compares are not permitted

### PR DESCRIPTION
**Pull request by seanyoung**
This solidity was not rejected and causes a panic in emit, because rationals should be compile-time constants.

```
function foo1(uint64 a, uint64 b) returns (bool) {
	return (a/b) >= 0.05;
}
```

Signed-off-by: Sean Young <sean@mess.org>

### CODE REVIEW

**Review by xermicus — CHANGES_REQUESTED** [review]
Thanks @seanyoung for so quickly tackling this!

But aren't there more operations? The following Solidity does crash the comiler as well:

```solidity
function foo1(uint64 a, uint64 b) public returns (bool) {
	return (a << b) >= 0.05;
}
```

**Review by xermicus — APPROVED** [review]

File: src/sema/expression.rs [hunk 1]
```diff
@@ -3905,11 +3937,19 @@ fn equal(
...
-    ))
+    );
+
+    if ty.is_rational() {
+        if let Err(diag) = eval_const_rational(&expr, ns) {
+            diagnostics.push(diag);
+        }
+    }
+
+    Ok(expr)
```
Comment (LucasSte) [inline src/sema/expression.rs:3952]: What about `NotEqual`?

File: tests/contract.rs [hunk 2]
```diff
@@ -33,18 +34,24 @@ fn contract_tests(file_path: &str, target: Target) -> io::Result<()> {
...
+
     for entry in read_dir(path)? {
         let path = entry?.path();
 
         if path.is_dir() {
             recurse_directory(path, target)?;
         } else if let Some(ext) = path.extension() {
             if ext.to_string_lossy() == "sol" {
-                parse_file(path, target)?;
+                entries.push(path);
```
Comment (LucasSte) [inline tests/contract.rs:46]: Why is this change necessary?

Comment (seanyoung) [reply tests/contract.rs:46]: I changed the test code so it uses `rayon` to use all available cores. This makes the test much faster

File: src/sema/expression.rs (hunk 1 above)
Comment (seanyoung) [reply src/sema/expression.rs:3952]: `NotEqual` is implement with `fn equal()` , so this works fine too. I've added a test for this.

File: tests/contract.rs (hunk 2 above)
Comment (LucasSte) [reply tests/contract.rs:46]: We already run Solana contracts, Substrate contracts and EVM contracts tests in parallel as they are different `#[test]` functions functions. Having another layer of parallelism here may slow down the tests (I have not tested though).

**Review by LucasSte — APPROVED** [review]
